### ISSUE TITLE
Merge Table Engine does not use the indexes of tables.

### ISSUE THREAD

**Issue by pwpow** [ClickHouse/ClickHouse#40014]
We use Merge Table Engine to merge a ReplicatedMergeTree Engine table and a normal view without Distributed Table Engine based on another ReplicatedMergeTree Engine table.

And we found that the Merge Table Engine does not use the indexes of view.

But if we use Distributed Table Engine to make these tables and views distributed, the Merge Table Engine will use their indexes.

So what is the difference between these two modes?

ClickHouse version: 21.9.2.17

**Comment by rkozlo** [comment]
Some more information and examples would be great

**Comment by den-crane** [comment]
Do you mean 

```
                                    ----> ReplicatedMergeTree  ( OK. here indexes are used. )
                                  /
                                 /
select --->>    EngineMerge  --- 
                                 \
                                  \
                                   ----> View ---> ReplicatedMergeTree ( Not OK. here indexes are not used. )
```

**Comment by pwpow** [comment]
@den-crane yep, and the indexes of two ReplicatedMergeTree are the same.

**Comment by pwpow** [comment]
The issue may be related to version 21.9.2.17, since this phenomenon cannot be repeated in version 20.3.8.53.
@rkozlo @den-crane

**Comment by pwpow** [comment]
The output of the indexes log is as follows:
![image](https://user-images.githubusercontent.com/35388969/183860086-cd43f0a4-9a57-4519-88d7-52db9380c50c.png)
![image](https://user-images.githubusercontent.com/35388969/183860144-0b8b5c4e-bbe3-49e7-9486-235f1f3b81a9.png)
![image](https://user-images.githubusercontent.com/35388969/183860235-13c437b9-aa37-4b4c-a1a5-61ec9b4f5864.png)
![image](https://user-images.githubusercontent.com/35388969/183860294-07cde058-7b08-489a-9ee4-45b5612104de.png)
![image](https://user-images.githubusercontent.com/35388969/183860501-1675e2a3-5d20-4bb2-ac45-79cc50f93e19.png)

If I create a same query just for the view, I can see 
![image](https://user-images.githubusercontent.com/35388969/183861703-9a50c22e-a4e9-44c9-9ce8-bb9e9602ca30.png)
![image](https://user-images.githubusercontent.com/35388969/183860780-4edeee4c-1e85-4fb4-9dc2-c128399411ce.png)
![image](https://user-images.githubusercontent.com/35388969/183860849-000d1d90-ef47-4863-87be-4d6fee941588.png)
![image](https://user-images.githubusercontent.com/35388969/183860909-6d326d94-0f96-4df9-8aec-d5673a96b2e7.png)
![image](https://user-images.githubusercontent.com/35388969/183860954-2b7836dd-16e2-4a06-9723-06db9cd83437.png)
![image](https://user-images.githubusercontent.com/35388969/183861015-be9ed78b-d6e8-4dac-9ded-192424d9f729.png)
@rkozlo @den-crane

**Comment by canhld94** [comment]
Reproduce with 

```sql
CREATE TABLE m0 (id UInt64) ENGINE=MergeTree ORDER BY id SETTINGS index_granularity = 1;
INSERT INTO m0 SELECT number FROM numbers(10);
CREATE TABLE m1 (id UInt64, s String) ENGINE=MergeTree ORDER BY id SETTINGS index_granularity = 1;
INSERT INTO m1 SELECT number, 'boo' FROM numbers(10);
CREATE VIEW m1v AS SELECT id FROM m1;

CREATE TABLE m2 (id UInt64) ENGINE=Merge(test,'m0|m1v');

SET max_bytes_to_read = 64;
SELECT * FROM m2 WHERE id > 1 AND id < 5;
```

My attempt to fix in #40233
